File: src/array/NSum.java
package array;


import java.util.*;

/**
 * NSum 系列
 * 给定一个整数数组 nums 和一个整数目标值 target，请你在该数组中找出 和为目标值 的那 两个 整数，并返回它们的数组下标。
 * 你可以假设每种输入只会对应一个答案。但是，数组中同一个元素在答案里不能重复出现。
 * 你可以按任意顺序返回答案。
 *
 * 输入：nums = [2,7,11,15], target = 9
 * 输出：[0,1]
 * 解释：因为 nums[0] + nums[1] == 9 ，返回 [0, 1] 。
 *
 * 1. 暴力法 -> hash法
 * 2. 排序后双指针, l<r
 */
public class NSum {
    public ArrayList<ArrayList<Integer>> twoSum(int[] nums, int start, int target) {
        ArrayList<ArrayList<Integer>> list = new ArrayList<ArrayList<Integer>>();

        if (nums.length == 0) return list;

        // sort nums
        Arrays.sort(nums);
        
        int n = nums.length;
        int l = start, r = n - 1;

        while (l < r) {
            int left = nums[l], right = nums[r];
            int sum = left + right;

            if (sum > target) {
                while (l < r && nums[r] == right) r--;
            } else if (sum < target) {
                while (l < r && nums[l] == left) l++;
            } else {
                ArrayList<Integer> tp = new ArrayList<Integer>();
                tp.add(left);
                tp.add(right);
                list.add(tp);   // {left, right}

                while (l < r && nums[r] == right) r--;
                while (l < r && nums[l] == left) l++;
            }
        }

        return list;
    }

    public ArrayList<ArrayList<Integer>> threeSum(int[] nums, int target) {
        ArrayList<ArrayList<Integer>> list = new ArrayList<ArrayList<Integer>>();

        if (nums.length < 3) return list;

        int n = nums.length;
        Arrays.sort(nums);

        for (int i = 0; i < n; i++) {
            ArrayList<ArrayList<Integer>> ans = twoSum(nums, i+1, target - nums[i]);  // &&& i+1 添加start防重

            for (int j = 0; j < ans.size(); j++) {
                ArrayList<Integer> tp = ans.get(j);
                tp.add(nums[i]);

                Collections.sort(tp);   // sort
                list.add(tp);
            }

            while (i < n - 1 && nums[i] == nums[i + 1]) i++;   // &&& 去重
        }

        return list;
    }

    public static void main(String[] args) {
        //int[] nums = {-2, 0, 1, 1, 2};
        int[] nums = {1,2,-2,-1};
        int target = 1;

        NSum o = new NSum();
        var res = o.threeSum(nums, target);
        System.out.println(res);
    }
}

File: src/array/nsum.cpp
/**
 * N数之和
 * 1. 两数之和
 * 15. 三数之和, a+b+c=0, 注意：答案中不可以包含重复的三元组。
 * 18. 四数之和，a+b+c+d=0, 注意：答案中不可以包含重复的三元组。
 *
 * 1. 前缀和
 * 排序后二分搜索；结果不能有重复
 * 15. 3Sum，调用2sum()
 * 18. nSum，递归nSum()
 */

#include <algorithm>
#include <stdio.h>
#include <vector>
#include <unordered_map>
#include "../util/myprint.h"

using namespace std;

/**
 * 二数之和索引，不能改变数组。引入map
 */
vector<int> twoSum(vector<int>& nums, int target) {
    // 元素 -> 索引
    unordered_map<int,int> map;

    // init
    for (int i=0; i<nums.size(); i++) {
        map.insert({nums[i], i});
    }

    for (int i=0; i<nums.size(); i++) {
        int other = target - nums[i];

        // &&& !=i
        if (map.find(other) != map.end() && map[other] != i) {
            return {i, map[other]};
        }
    }

    return {};
}

/**
 * 2Sum 之和，需要先排序
 */
vector<vector<int>> twoSumTarget(vector<int>& nums, int target) {
    vector<vector<int>> res;

    // 排序
    sort(nums.begin(), nums.end());

    int lo = 0, hi = nums.size() - 1; // &&& =0

    // 二分搜索
    while (lo  < hi) {
        int left = nums[lo], right = nums[hi];      // &&& 赋值
        int sum = left + right;

        if (sum > target) {
            // 去重
            while (lo < hi && nums[hi] == right)  hi--;     // &&& l<r
        } else if (sum < target) {
            // 去重
            while (lo < hi && nums[lo] == left)  lo++;
        } else {
            res.push_back({left, right});

            // &&& 去重
            while (lo < hi && nums[hi] == right)  hi--;
            while (lo < hi && nums[lo] == left)  lo++;
        }
    }

    return res;
}


/**
 * 三数之和，call twoSumTarget
 */
vector<vector<int>> threeSumTarget(vector<int>& nums, int target) {
    vector<vector<int>> res;
    int n = nums.size();

    if (n == 0) {
        return res;
    }

    sort(nums.begin(), nums.end());

    for (int i=0; i<n; i++) {
        auto tuples = TwoSumTarget(nums, i+1, target - nums[i]);   // &&&剩余两数之和

        // 补全
        for (auto &tuple : tuples) {
            tuple.push_back(nums[i]);

            res.push_back(tuple);
        }

        // 去重
        while (i < n-1 && nums[i] == nums[i+1]) i++;
    }

    return res;
}


vector<vector<int>> threeSumTarget1(vector<int>& nums, int target) {
    sort(nums.begin(), nums.end());

    return nSumTarget(nums, 3, 0, target);
}

int main(int argc, char **argv) {
    // 2Sum
    vector<int> v0{3, 2, 4};
    auto ret = twoSum(v0, 6);
    show(ret);

    // 2Sum
    vector<int> v{1, 1, 1, 2, 2, 3, 3};
    auto res = twoSumTarget(v, 4);
    showV(res);

    // 3Sum
    vector<int> v1{-1, 0, 1, 2, -1, -4};
    res = threeSumTarget(v1, 0);
    showV(res);

    // 3Sum
    res = threeSumTarget1(v1, 0);
    showV(res);

    return 0;
}